lib/vector-tap/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+            component_type_label: "component_type".green(),
+        }
+    }
+
+    pub fn format<'a>(
+        &self,
+        component_id: &str,
+        component_kind: &str,
+        component_type: &str,
+        event: &'a str,
+    ) -> Cow<'a, str> {
+        if self.meta {
+            match self.format {
+                TapEncodingFormat::Json => format!(
+                    r#"{{"{}":"{}","{}":"{}","{}":"{}","event":{}}}"#,
+                    self.component_id_label,
+                    component_id.green(),
+                    self.component_kind_label,
+                    component_kind.green(),
+                    self.component_type_label,
+                    component_type.green(),
+                    event
+                )
+                    .into(),
+                TapEncodingFormat::Yaml => {
+                    let mut value: BTreeMap<String, serde_yaml::Value> = BTreeMap::new();
+                    value.insert("event".to_string(), serde_yaml::from_str(event).unwrap());
+                    // We interpolate to include component_id rather than
+                    // include it in the map to correctly preserve color
+                    // formatting
+                    format!(
+                        "{}{}: {}\n{}: {}\n{}: {}\n",
+                        serde_yaml::to_string(&value).unwrap(),
+                        self.component_id_label,
+                        component_id.green(),
+                        self.component_kind_label,
+                        component_kind.green(),
+                        self.component_type_label,
+                        component_type.green()
+                    )
+                        .into()
+                }
+                TapEncodingFormat::Logfmt => format!(
+                    "{}={} {}={} {}={} {}",
+                    self.component_id_label,
+                    component_id.green(),
+                    self.component_kind_label,
+                    component_kind.green(),
+                    self.component_type_label,
+                    component_type.green(),
+                    event
+                )
+                    .into(),
+            }
+        } else {
+            event.into()
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum OutputChannel {
+    Stdout(EventFormatter),
+    AsyncChannel(tokio_mpsc::Sender<Vec<GraphQLTapOutputEvents>>),
+}
+
+/// Error type for DNS message parsing
+#[derive(Debug, Snafu)]
+pub enum TapExecutorError {
+    #[snafu(display("[tap] Couldn't connect to API via WebSockets"))]
+    ConnectionFailure,
+    GraphQLError,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn exec_tap(
+    url: &Url,
+    interval: i64,
+    limit: i64,
+    duration_ms: Option<u64>,
+    input_patterns: Vec<String>,
+    output_patterns: Vec<String>,
+    format: TapEncodingFormat,
+    formatter: &EventFormatter,
+    quiet: bool,
+) -> Result<(), TapExecutorError> {
```

Comment:
With this many parameters, in particular with several being identical types, I would recommend making a `struct` to hold at least some of them and move the implementation to a member function.